File: backend/src/main/kotlin/com/example/backend/services/TypeSortieService.kt
package com.example.backend.services

import com.example.backend.dtos.SortieDetailDto
import com.example.backend.dtos.TypeSortieDto
import com.example.backend.models.ProduitDetail
import com.example.backend.models.SortieStock
import com.example.backend.models.TypeSortie
import com.example.backend.repositories.*
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.stereotype.Service
import java.time.LocalDateTime
import kotlin.jvm.optionals.getOrNull

@Service
class TypeSortieService(
  private val produitDetailRepository: ProduitDetailRepository,
  private val sortieStockRepository: SortieStockRepository,
  private val typeSortieRepository: TypeSortieRepository,
  private val enRayonRepository: EnRayonRepository,
  private val produitRepository: ProduitRepository
) {

  fun getTypeSortiePageable(
    nom: String?,
    pageable: Pageable
  ): Page<Map<String, Any?>> {
    val specification = TypeSortieRepository.filterTypeSortie(nom)
    return typeSortieRepository.findAll(specification, pageable)
      .map { typeSortie ->
        mapOf(
          "id" to typeSortie.id,
          "nom" to typeSortie.nom
        )
      }
  }


  fun addTypeSortiel(sortie: TypeSortieDto): TypeSortie {
    if (sortie.id==0){
      var newTypeSortie = TypeSortie().apply {
        this.nom = sortie.nom
        this.description = sortie.description
      }
      newTypeSortie = typeSortieRepository.save(newTypeSortie)
      return newTypeSortie
    }
    else {
      val typeSortie = typeSortieRepository.findById(sortie.id!!.toInt()).getOrNull()
      typeSortie!!.nom = sortie.nom
      typeSortie.description = sortie.description
      return typeSortieRepository.save(typeSortie!!)
    }
  }
}

File: backend_py/pharmaPython/app/services/type_sortie_service.py
# services/type_sortie_service.py
from sqlalchemy.orm import Session

from app.repositories.type_sortie_repository import TypeSortieRepository


class TypeSortieService:
  def __init__(self, db: Session):
    self.db = db
    self.type_sortie_repo = TypeSortieRepository(db)

  def get_type_sortie_pageable(self, nom: str, page: int, size: int):
    rows, total = self.type_sortie_repo.filter_type_sortie(nom, page, size)
    return {"content": [{"id": r.id, "nom": r.nom} for r in rows], "totalElements": total}

  def add_type_sortie(self, dto):
    if dto.id == 0:
      new = self.type_sortie_repo.save({"nom": dto.nom, "description": dto.description})
      return new
    else:
      ts = self.type_sortie_repo.find_by_id(dto.id)
      ts.nom = dto.nom
      ts.description = dto.description
      return self.type_sortie_repo.save(ts)
